### PR TITLE
Minor GKS formatting changes and addition of gnomAD flags to annotation

### PR DESCRIPTION
**Pull request by theferrit32**
Modifications:
- change `popFreqID` to `popFreqId`
- change subcohort ids to use a `.` delimiter instead of `,` to match the popmax syntax

Additions:
- add qcFilters
- add lowComplexityRegion (lcr)
- add allele balance flagged count (>0.9 in the table) as `heterozygousSkewedAlleleCount`

One point is that the allele balance code will fail if the number of entries or the bins in the array `ht.qual_hists.ab_hist_alt.bin_freq` changes. That field is directly used to drive the the UI.

EDIT:
Additional changes:
- add hemizygote counts to variant and subcohorts
- if no popmax faf exists, sets the popmaxFAF95 object to null instead of setting the freq to 0 and having null or other string in other fields (@matren395)
- rename `popmaxFAF95` to `grpMaxFAF95`, `popFreqId` to `grpFreqId` (@matren395)

### CODE REVIEW

**Comment by matren395** [comment]
Hey Kyle, could we roll the change to "popFreqId" not returning "None" but instead a string here as well? I had created my own PR for that and the "popFreqID" fix, but it looks like this already contains the latter and would be easy enough to host the former as well

File: gnomad/utils/annotations.py
Comment (klaricch) [inline gnomad/utils/annotations.py:2304]:
```suggestion
    # The bin_freq should be in order but we can verify the order from bin_edges.
```

File: gnomad/utils/annotations.py
Comment (klaricch) [inline gnomad/utils/annotations.py:2318]:
```suggestion
    # Create ancillaryResults for additional frequency and popMaxFAF95 information.
```

File: gnomad/utils/annotations.py
Comment (klaricch) [inline gnomad/utils/annotations.py:2304]:
```suggestion
    # The bin_freq should be in order but we can verify the order from bin_edges.
```

File: gnomad/utils/annotations.py
Comment (klaricch) [inline gnomad/utils/annotations.py:2327]: so this is just the count of how many genotypes were flagged?

File: gnomad/utils/annotations.py
Comment (klaricch) [inline gnomad/utils/annotations.py:2307]: what does tpl stand for?

**Review by klaricch — CHANGES_REQUESTED** [review]

File: gnomad/utils/annotations.py
Comment (theferrit32) [reply gnomad/utils/annotations.py:2325]: "tuple"

File: gnomad/utils/annotations.py
Comment (theferrit32) [reply gnomad/utils/annotations.py:2345]: Yes the count of those over the 0.9 threshold. Let me know if this seems correct or if there are any weird cases that might need to be handled differently

File: gnomad/utils/annotations.py
Comment (theferrit32) [reply gnomad/utils/annotations.py]: I've added it just now based on the discussion, but it can be changed if they decide on something else

**Comment by theferrit32** [comment]
One more pending change is to add `count of hemizygotes`. Which needs to be computed, it is not explicitly in a field on the table.

@larrybabb @ahwagner  Was there an idea to move some of these fields out of `ancillaryResults`?

see: https://github.com/ga4gh/va-spec/pull/119#issuecomment-1751097147

I think some of that might have been related to my confusion. I'm not sure the allele balance flagged actually are directly related to the popmax FAF 95. Maybe someone on the gnomad team can clarify.

File: gnomad/utils/annotations.py
Comment (klaricch) [inline gnomad/utils/annotations.py:2326]:
```suggestion
            lambda x: x[1],
```

File: gnomad/utils/annotations.py
Comment (klaricch) [inline gnomad/utils/annotations.py:2332]:
```suggestion
                key=lambda x: x[0],
```

File: gnomad/utils/annotations.py
Comment (klaricch) [inline gnomad/utils/annotations.py:2352]: looks like we're no longer going with the NO_COHORTS option so all of popMaxFAF95 should be null in these cases

File: gnomad/utils/annotations.py
Comment (klaricch) [inline gnomad/utils/annotations.py:2346]: regarding your comment, i'm not aware of any way this is related to faf

**Review by klaricch — CHANGES_REQUESTED** [review]

**Comment by matren395** [comment]
commit to roll in 1) nullable popMaxFAF95 and 2) change all references of 'pop' returned to 'grp' when relevant.

File: gnomad/utils/annotations.py
Comment (matren395) [inline gnomad/utils/annotations.py:2372]: @klaricch this is the added code (in this region) to fix their new schema for accepting NONE types - lmk if this looks right

**Comment by theferrit32** [comment]
Waiting for final modifications to upstream schema. There is one more discussion tomorrow. The changes will be relatively easy to implement after that.

See https://github.com/ga4gh/va-spec/pull/121

**Comment by theferrit32** [comment]
Adding commits and force pushing a branch rebased from main

File: gnomad/resources/grch38/gnomad.py
Comment (matren395) [inline gnomad/resources/grch38/gnomad.py:466]: remind me, is the rbind code the same from before?

File: gnomad/resources/grch38/gnomad.py
Comment (matren395) [inline gnomad/resources/grch38/gnomad.py:463]: Remind me, the I'm having a hard time parsing the Schema. 'GrpMaxFAF95' is required to exist but can just return values of None and None within it?

File: gnomad/utils/annotations.py
Comment (matren395) [inline gnomad/utils/annotations.py:2252]: looks like a bit of code - does this speed up performance?

File: gnomad/utils/annotations.py
Comment (matren395) [inline gnomad/utils/annotations.py:2350]: is this the same code that starts at line #2279? This looks like the simpler version here, but I wanna make sure that I'm not just turned around

**Review by matren395 — COMMENTED** [review]
might take another run at the actual code in annotations.py later today

File: gnomad/resources/grch38/gnomad.py
Comment (theferrit32) [reply gnomad/resources/grch38/gnomad.py:466]: @matren395 yes this is just a copy-paste of the rbind code you wrote. Just moved it to a function so I could swap between v3 and v4 strategies.

File: gnomad/resources/grch38/gnomad.py
Comment (theferrit32) [reply gnomad/resources/grch38/gnomad.py:463]: It's not required to exist anymore, instead of making it required and allowing None, they just made it optional.

https://github.com/ga4gh/va-spec/blob/1.0-alpha/schema/cohortAlleleFreq.json#L341-L359

File: gnomad/utils/annotations.py
Comment (theferrit32) [reply gnomad/utils/annotations.py:2252]: I think the reason I did this is so all the `freq` keys are created in this inner function instead of where the function is called. They are all based on the group_id without sex information. Previously it was sending both a group_id including sex, and the sex label itself. And now that we are also pulling out the `hemizogytes`, we need the group_id without sex in order to make the key with XY.

File: gnomad/utils/annotations.py
Comment (theferrit32) [reply gnomad/utils/annotations.py:2350]: Essentially yes, the reason this is simpler is because it's the top level group so the key is always just `XY_adj`. In the inner function it needs to check whether the current group is sexed or not, and if not, query the `freq` for the sexed sub-group.

File: gnomad/resources/grch38/gnomad.py
Comment (theferrit32) [reply gnomad/resources/grch38/gnomad.py:549]: @klaricch is there just exomes coverage available in v4? That's all I see in the gnomad downloads page and in the `coverage` function. If coverage info is requested for a v4 genomes table, what should the behavior be?

**Comment by theferrit32** [comment]
Latest changeset: https://github.com/theferrit32/gnomad_methods/compare/374112031d6b1178e40cfc23161a19b82b9d5693..74d869e022933612e566755f42d8dd2e0445e114

- fixes fafmax for exomes
- uses v4 exomes coverage (for both v4 genomes and v4 exomes, is this correct? or should genomes tables not be given any coverage info?)
- replaces AF = None (AC/AN = 0/0) with 0. (the jsonschema requires number, can't be null)

File: gnomad/resources/grch38/gnomad.py
Comment (matren395) [reply gnomad/resources/grch38/gnomad.py]: ah! gnomAD v4 genomes are the same as gnomAD v3 genomes , so the coverage ought to be the same for v4 genomes as for v3 genomes.  This might require some extra logic with the user selecting a data type - I forget if that is already requested

File: gnomad/resources/grch38/gnomad.py
Comment (matren395) [reply gnomad/resources/grch38/gnomad.py:466]: 👍

File: gnomad/resources/grch38/gnomad.py
Comment (matren395) [reply gnomad/resources/grch38/gnomad.py:463]: The way I'm always reading it, in this code, it looks like GrpMaxFAF95 always exists but everything it's populated with is allowed to be None. 1) Am I reading this implementation right 2) this implementation is allowed, right? Sorry if this is v base level

File: gnomad/utils/annotations.py
Comment (matren395) [reply gnomad/utils/annotations.py:2354]: Can we remove the code at line #2279 then? Again sorry if I'm parsing this wrong

File: gnomad/utils/annotations.py
Comment (matren395) [reply gnomad/utils/annotations.py:2252]: 👍

File: gnomad/utils/annotations.py
Comment (matren395) [reply gnomad/utils/annotations.py:2354]: adding a new review, pinging this again if we can remove the code at line #2279-2290 given this code

File: gnomad/resources/grch38/gnomad.py
Comment (matren395) [inline gnomad/resources/grch38/gnomad.py:560]: this is tricky, since v4 has genomes (from v3 but still) and exomes as well. could include the Data_Type parameter and say: 

if data_type==exomes: 
     if high_level_version == "v3": ERROR
     elif high_level_version == "v4": ....coverage("exomes")...
elif data_type==genomes:
     ...coverage("genomes")

but I'm not sure if the information in v4 genomes is different than v3 genomes ? I know it's the same samples but idk if it's identical

**Review by matren395 — COMMENTED** [review]
notes! lmk if you wanna chat abt the PR next week

File: gnomad/resources/grch38/gnomad.py
Comment (klaricch) [reply gnomad/resources/grch38/gnomad.py:560]: just summing up coverage comments here:
I put in a fix to the coverage function while working on constraint so that it no longer tries to pull coverage versions that don't exist ( so you no longer have to assign "coverage_version" to avoid errors). v4 genomes is v3 genomes, so there are no new genome coverage files for v4. You can just pull call the coverage function and specify data_type and don't need to specify version. So for v4, using a datatype of either exomes or genomes will be fine since genomes will just pull the v3 coverage data. For v3, using a datatype of genomes is fine, but should error out if a user tries to pull exomes for v3.

"over_20" should also be added to the coverage annotation to be consistent with what's shown on the browser

File: gnomad/utils/annotations.py
Comment (theferrit32) [reply gnomad/utils/annotations.py:2354]: @matren395 I don't think the lines 2279-2289 can be removed or combined here, it adds logic to look up the specific population group. It's possible both 2348-2350 and 2279-2289 could be pulled out into a function but it would probably be more lines and more complicated

File: gnomad/resources/grch38/gnomad.py
Comment (matren395) [reply gnomad/resources/grch38/gnomad.py:463]: ah okay, makes sense ! thanks for the answer

File: gnomad/utils/annotations.py
Comment (matren395) [reply gnomad/utils/annotations.py:2354]: noted! thanks for the explanation

File: gnomad/utils/annotations.py
Comment (matren395) [reply gnomad/utils/annotations.py:2358]: makes sense, looks good!

File: gnomad/resources/grch38/gnomad.py
Comment (matren395) [reply gnomad/resources/grch38/gnomad.py]: thinks this was resolved someplace else, function structuring here makes sense

File: gnomad/resources/grch38/gnomad.py
Comment (klaricch) [reply gnomad/resources/grch38/gnomad.py:560]: @theferrit32 it looks like coverage is the last item that still needs changes, can you edit the functions to address the above comment regarding coverage?

Comment (theferrit32) [reply gnomad/resources/grch38/gnomad.py:560]: @klaricch I'm about to push a change to handle that based on your comment above.

For the "over_20" comment, are you suggesting to add the `.over_20` field from the coverage_ht to the annotation output in addition to the `meanDepth` annotation?

@ahwagner @larrybabb do we want to add a field to include the `Fraction of individuals with >20x coverage` field added to the browser UI?

Comment (klaricch) [reply gnomad/resources/grch38/gnomad.py:560]: yeah over_20 in additions to meanDepth

Comment (larrybabb) [reply gnomad/resources/grch38/gnomad.py:560]: @theferrit32 I'm fine with adding the two new fields, but I defer to @ahwagner for the final say. I'll make sure he's seeing these questions.

Comment (ahwagner) [reply gnomad/resources/grch38/gnomad.py:560]: @theferrit32 to make sure I'm following this proposal, is this an addition of `over_20` to the `caf.qualityMeasures` object? And no other additions? If so, this seems like the appropriate place to add the field and I have no issue with it so long as there is a PR to update the [model](https://github.com/ga4gh/va-spec/blob/1.0-alpha/schema/cohortAlleleFreq.yaml) and [examples](https://github.com/ga4gh/va-spec/blob/1.0-alpha/docs/Modeling/CohortFrequency/simple_result_example.yaml) at VA-Spec.

Comment (theferrit32) [reply gnomad/resources/grch38/gnomad.py:560]: @ahwagner https://github.com/ga4gh/va-spec/pull/124

File: gnomad/resources/grch38/gnomad.py
Comment (theferrit32) [inline gnomad/resources/grch38/gnomad.py:562]: @klaricch this will use `genomes` version `3.0.1` coverage table for v3 and v4 genomes tables, and `exomes` version `4.0` coverage table for v4 exomes tables. If the input is a v3 table with `data_type=exomes`, line 562 will throw an exception `KeyError: '3.0.1'`, which is expected.

File: gnomad/sample_qc/ancestry.py
Comment (theferrit32) [inline gnomad/sample_qc/ancestry.py:53]: This is resulting in a python error when merging into main, because it was added separately in that branch, on a different line

https://github.com/broadinstitute/gnomad_methods/blob/main/gnomad/sample_qc/ancestry.py#L35

**Comment by theferrit32** [comment]
Rebasing from main to fix https://github.com/broadinstitute/gnomad_methods/pull/617#pullrequestreview-1816172266

**Review by matren395 — COMMENTED** [review]
Coveraged looks covered!

**Comment by theferrit32** [comment]
I think @klaricch has to submit a final review approval since she had a prior request for changes

File: gnomad/resources/grch38/gnomad.py
Comment (matren395) [inline gnomad/resources/grch38/gnomad.py:537]:
```suggestion
    coverage_version_3_0_1 = "3.0.1"  # v4 genomes coverage
```

File: gnomad/resources/grch38/gnomad.py
Comment (matren395) [inline gnomad/resources/grch38/gnomad.py:543]:
```suggestion
    if data_type == "genomes":
```

File: gnomad/resources/grch38/gnomad.py
Comment (matren395) [inline gnomad/resources/grch38/gnomad.py:549]:
```suggestion
    else:
       coverage_version = coverage_version_4_0
```

File: gnomad/resources/grch38/gnomad.py
Comment (matren395) [inline gnomad/resources/grch38/gnomad.py:552]:
```suggestion
            "gnomad_gks() is currently only implemented for gnomAD v4."
```

File: gnomad/resources/grch38/gnomad.py
Comment (matren395) [inline gnomad/resources/grch38/gnomad.py:550]:
```suggestion
    
    if high_level_version != "v4":
```

File: gnomad/resources/grch38/gnomad.py
Comment (matren395) [inline gnomad/resources/grch38/gnomad.py:589]:
```suggestion
    ht = ht.annotate(lcr=ht.region_flags.lcr)
```

File: gnomad/resources/grch38/gnomad.py
Comment (matren395) [inline gnomad/resources/grch38/gnomad.py:542]:
```suggestion
```

**Review by matren395 — COMMENTED** [review]
An effort to clean up and show that we only support v4 versions - let me know if I'm missing anything , might take a finer pass at this tomorrow

File: gnomad/resources/grch38/gnomad.py
Comment (matren395) [inline gnomad/resources/grch38/gnomad.py:600]:
```suggestion
    ht = ht.annotate(ab_hist_alt=ht.histograms.qual_hists.ab_hist_alt)

    ht = add_grpMaxFAF95_v4(ht)
```

Comment (matren395) [reply gnomad/resources/grch38/gnomad.py:600]: appropriate change if fully not supporting v3 at all - may need beautified with some more comments or documentation around here, or running black on it again

**Review by matren395 — COMMENTED** [review]
some more version-relevant changes if we are removing support for v3

File: gnomad/utils/annotations.py
Comment (theferrit32) [reply gnomad/utils/annotations.py:2246]: Yeah that was used when I was initially implementing by using a VCF export/import strategy. I will delete now.

Comment (theferrit32) [reply gnomad/utils/annotations.py:2165]: Deleted

File: gnomad/resources/grch38/gnomad.py
Comment (matren395) [inline gnomad/resources/grch38/gnomad.py:508]: Just occurred to me - would make more sense to move this to the start of the function (and accompanying high_level_version logic) before reading in any tables. 

@klaricch your other PR has already been merged and we can leave it up to the other functions to fail for bad keys, yes ? We don't want to check here that like 'version' is in the list of valid versions or anything, yeah ?

**Review by matren395 — COMMENTED** [review]
only one thought to make the logic more logical

**Review by klaricch — CHANGES_REQUESTED** [review]

File: gnomad/resources/grch38/gnomad.py
Comment (theferrit32) [reply gnomad/resources/grch38/gnomad.py:508]: I will move the high_level_version stuff up above the first table read

**Review by klaricch — APPROVED** [review]

**Pull request merged**
